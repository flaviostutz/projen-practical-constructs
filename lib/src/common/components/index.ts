export * from './base-tooling';
export * from './common-target-type';
export * from './common-targets';
export * from './release-tasks';
export * from './publish';
