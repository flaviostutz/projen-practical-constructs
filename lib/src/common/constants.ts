export const PROJEN_VERSION = '0.91.13';
export const NODE_VERSION = '20.16.0';
export const MONOTAG_VERSION = '1.23.0';
