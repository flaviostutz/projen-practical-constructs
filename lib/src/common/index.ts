export * from './components';
export * from './files';
export * from './types';
