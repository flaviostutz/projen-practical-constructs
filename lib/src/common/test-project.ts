import { mkdtempSync } from 'node:fs';
import { tmpdir } from 'node:os';
// eslint-disable-next-line unicorn/import-style
import { join } from 'node:path';

import { Project } from 'projen';

export class TestProject extends Project {
  constructor() {
    const outdir = mkdtempSync(join(tmpdir(), 'test-'));
    super({ name: 'test', outdir });
  }
}
