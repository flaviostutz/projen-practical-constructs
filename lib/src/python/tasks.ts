import { Project, Task } from 'projen';

export interface TaskOptionsTarget {
  /**
   * Path to the python virtual environment directory
   * used in this project
   * @default .venv
   */
  readonly venvPath?: string;
  /**
   * Existing task to attach new tasks to.
   * It will be included as "spawn" tasks in new steps
   */
  readonly attachTasksTo?: string;
}

export interface TaskOptions {
  /**
   * Path to the python virtual environment directory
   * used in this project
   */
  readonly venvPath: string;
  /**
   * Existing task to attach new tasks to.
   * It will be included as "spawn" tasks in new steps
   */
  readonly attachTasksTo?: string;
}

export interface TaskOptionsWithFix extends TaskOptions {
  /**
   * Attach task to fix
   */
  readonly attachFixTasksTo?: string;
}

export const addTaskToParent = (project: Project, task: Task, parentTaskName?: string): void => {
  if (!parentTaskName) {
    return;
  }
  const attachTask = project.tasks.tryFind(parentTaskName);
  if (!attachTask) {
    throw new Error(`'${parentTaskName}' task not found`);
  }
  attachTask.spawn(task);
};
