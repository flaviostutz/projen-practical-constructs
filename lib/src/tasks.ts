import { Project, Task } from 'projen';

export interface TaskOptions {
  /**
   * Path to the python virtual environment directory
   * used in this project
   * @default .venv
   */
  readonly venvPath?: string;
  /**
   * Existing task to attach new tasks to.
   * It will be included as "spawn" tasks in new steps
   */
  readonly attachTasksTo?: string;
}

export const addTaskToParent = (project: Project, task: Task, parentTaskName?: string): void => {
  if (!parentTaskName) {
    return;
  }
  const attachTask = project.tasks.tryFind(parentTaskName);
  if (!attachTask) {
    throw new Error(`'${parentTaskName}' task not found`);
  }
  attachTask.spawn(task);
};
