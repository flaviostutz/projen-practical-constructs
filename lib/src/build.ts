/* eslint-disable no-restricted-syntax */
/* eslint-disable no-new */
import { Component, Project } from 'projen';

import { Pip, PipOptions } from './components/pip';

/**
 * Python project build configurations
 */
export class BuildTarget extends Component {
  constructor(project: Project, opts: BuildOptions) {
    super(project);

    if (!opts.venvPath) {
      throw new Error('venvPath is required');
    }
    if (!opts.lockFile) {
      throw new Error('lockFile is required');
    }

    project.tasks.addTask('build', {
      description: `Build project (install deps, compile and package)`,
    });

    new Pip(project, { ...opts, attachTasksTo: 'build' });

    project.addGitIgnore('.ipynb_checkpoints');
  }
}

export interface BuildOptions extends PipOptions {}
