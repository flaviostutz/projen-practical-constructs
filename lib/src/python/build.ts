/* eslint-disable no-restricted-syntax */
/* eslint-disable no-new */
import { Component, Project } from 'projen';

import { Pip, PipOptions } from './components/pip';
import { Package, PackageOptions } from './components/package';
import { PythonVersionFile } from './files';
import { TaskOptions } from './tasks';

/**
 * Python project build configurations
 */
export class BuildTarget extends Component {
  constructor(project: Project, taskOpts: TaskOptions, opts?: Build0Options) {
    super(project);

    project.tasks.addTask('build', {
      description: `Build project (install deps, compile and package)`,
    });

    // create package such as pyproject.toml, LICENSE etc
    new Package(project, opts?.package);

    // create pip resources such as tasks etc
    new Pip(project, taskOpts, opts?.pip);

    // create python-version (can be used by pyenv etc)
    const requiresPython = opts?.package?.requiresPython ?? '3.12';
    const pythonVersion = requiresPython.replace(/[^0-9.]/g, '');
    new PythonVersionFile(project, { pythonVersion });

    project.addGitIgnore('.ipynb_checkpoints');
  }
}

// This interface is not called BuildOptions because JSII doesn't allow the usage of name "build" on members
export interface Build0Options {
  readonly pip?: PipOptions;
  readonly package?: PackageOptions;
}
