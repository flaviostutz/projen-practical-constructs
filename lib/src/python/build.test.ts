/* eslint-disable no-new */
import { Testing } from 'projen';

import { TestProject } from '../common/test-project';

import { BuildTarget } from './build';
import { TaskOptions } from './tasks';

describe('BuildTarget', () => {
  test('BuildTarget is synthesized correctly', () => {
    const project = new TestProject();
    const taskOpts: TaskOptions = { venvPath: '.venv' };
    project.removeTask('build');
    new BuildTarget(project, taskOpts);
    const output = Testing.synth(project);
    expect(output).toMatchSnapshot();
  });

  test('BuildTarget synthesizes correctly', () => {
    const project = new TestProject();
    const taskOpts: TaskOptions = { venvPath: '.venv' };
    project.removeTask('build');
    new BuildTarget(project, taskOpts, {
      package: {
        version: '1.2.3',
      },
      pip: {
        pythonExec: 'python3',
      },
    });
    const output = Testing.synth(project);
    expect(output).toMatchSnapshot();
  });
});
