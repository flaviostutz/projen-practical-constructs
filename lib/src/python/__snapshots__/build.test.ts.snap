// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildTarget BuildTarget is synthesized correctly 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/pyproject.toml
.venv/
*.egg-info
build
*.pyc
__pycache__
.cache
!/.python-version
.ipynb_checkpoints
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "pip-tools",
        "type": "devenv",
        "version": "==7.4.1",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      ".python-version",
      "pyproject.toml",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Build project (install deps, compile and package)",
        "name": "build",
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "install": {
        "description": "Install dependencies from constraints.txt",
        "name": "install",
        "steps": [
          {
            "exec": ".venv/bin/pip install --require-virtualenv -c constraints.txt",
          },
        ],
      },
      "install-dev": {
        "description": "Install dependencies from constraints-dev.txt",
        "name": "install-dev",
        "steps": [
          {
            "exec": ".venv/bin/pip install --require-virtualenv -c constraints-dev.txt .[dev]",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "prepare-venv": {
        "description": "Create python virtual environment in .venv",
        "name": "prepare-venv",
        "steps": [
          {
            "exec": "python -m venv .venv",
          },
          {
            "exec": ".venv/bin/pip install pip-tools==7.4.1",
          },
        ],
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
      "update-lockfile": {
        "description": "Update lock file (constraints.txt) according to pyproject.toml",
        "name": "update-lockfile",
        "steps": [
          {
            "say": "Prepare venv",
            "spawn": "prepare-venv",
          },
          {
            "exec": ".venv/bin/pip-compile --all-build-deps --output-file=constraints.txt pyproject.toml",
            "say": "Updating lock file (runtime)",
          },
          {
            "exec": ".venv/bin/pip-compile --all-build-deps --extra dev --strip-extras --output-file=constraints-dev.txt pyproject.toml",
            "say": "Updating lock file (dev)",
          },
        ],
      },
    },
  },
  ".python-version": "3.12",
  "pyproject.toml": "[build-system]
requires = [ "setuptools" ]
build-backend = "setuptools.build_meta"

[project]
name = "test"
version = "0.0.1"
description = ""
readme = ""
requires-python = ">=3.12"
license = { file = "LICENSE" }
keywords = [ ]
dependencies = [ "pip-tools==7.4.1" ]

[project.optional-dependencies]
dev = [ ]

[tool.setuptools.package-data]
test = [ "py.typed" ]
",
}
`;

exports[`BuildTarget BuildTarget synthesizes correctly 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

* text=auto eol=lf",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/pyproject.toml
.venv/
*.egg-info
build
*.pyc
__pycache__
.cache
!/.python-version
.ipynb_checkpoints
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "dependencies": [
      {
        "name": "pip-tools",
        "type": "devenv",
        "version": "==7.4.1",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "files": [
      ".gitattributes",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      ".python-version",
      "pyproject.toml",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Build project (install deps, compile and package)",
        "name": "build",
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "install": {
        "description": "Install dependencies from constraints.txt",
        "name": "install",
        "steps": [
          {
            "exec": ".venv/bin/pip install --require-virtualenv -c constraints.txt",
          },
        ],
      },
      "install-dev": {
        "description": "Install dependencies from constraints-dev.txt",
        "name": "install-dev",
        "steps": [
          {
            "exec": ".venv/bin/pip install --require-virtualenv -c constraints-dev.txt .[dev]",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "prepare-venv": {
        "description": "Create python virtual environment in .venv",
        "name": "prepare-venv",
        "steps": [
          {
            "exec": "python3 -m venv .venv",
          },
          {
            "exec": ".venv/bin/pip install pip-tools==7.4.1",
          },
        ],
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
      "update-lockfile": {
        "description": "Update lock file (constraints.txt) according to pyproject.toml",
        "name": "update-lockfile",
        "steps": [
          {
            "say": "Prepare venv",
            "spawn": "prepare-venv",
          },
          {
            "exec": ".venv/bin/pip-compile --all-build-deps --output-file=constraints.txt pyproject.toml",
            "say": "Updating lock file (runtime)",
          },
          {
            "exec": ".venv/bin/pip-compile --all-build-deps --extra dev --strip-extras --output-file=constraints-dev.txt pyproject.toml",
            "say": "Updating lock file (dev)",
          },
        ],
      },
    },
  },
  ".python-version": "3.12",
  "pyproject.toml": "[build-system]
requires = [ "setuptools" ]
build-backend = "setuptools.build_meta"

[project]
name = "test"
version = "1.2.3"
description = ""
readme = ""
requires-python = ">=3.12"
license = { file = "LICENSE" }
keywords = [ ]
dependencies = [ "pip-tools==7.4.1" ]

[project.optional-dependencies]
dev = [ ]

[tool.setuptools.package-data]
test = [ "py.typed" ]
",
}
`;
